tests/framework/project/test_logging.py

```diff
@@ -23,9 +23,22 @@ def default_logging_config():
     return logging_config
 
 
+@pytest.fixture
+def default_logging_config_with_project():
+    logging_config = {
+        "version": 1,
+        "disable_existing_loggers": False,
+        "handlers": {
+            "rich": {"class": "kedro.logging.RichHandler", "rich_tracebacks": True}
+        },
+        "loggers": {"kedro": {"level": "INFO"}, "test_project": {"level": "INFO"}},
+        "root": {"handlers": ["rich"]},
+    }
+    return logging_config
+
+
 @pytest.fixture(autouse=True)
 def reset_logging(default_logging_config):
-    yield
```

Comment:
Unsure why yield statement was introduced as it prevents `LOGGING` from being reset between logging tests

Reply:
Ah, I see -  I've changed it because `LOGGING.data` perpetuates from tests that are run on different mock projects before this particular suite of tests, and for each mock project a corresponding logger is now added to the default logger - which would cause `test_default_logging_config` to fail. These project loggers cannot be explicitly defined in `default_logging_config` as it is dependent on the order of the tests run. 

Are there any implications to the tests by having the logs reset before them as opposed to after? As I understand it the ordering of operations is the same for the most part.

Reply:
Ideally the order in which the tests are run shouldn't matter and they should pass. If that's not the case then data cleanup isn't done properly after each test. If `LOGGING.data` is causing issues, I'd assume we have to reset that after each test.

Reply:
I agree @merelcht , the only issue is that `LOGGING.data` isn't being reset after other tests modify it in test suites that are run before this file gets run - running this test file alone behaves as expected as there are no pre-existing modifications to `LOGGING.data` and it is reset after every test. The problem arises when running this test suite after the tests in `test_startup.py` (and some other files) - `LOGGING.data` is not reset after those tests, and so the first test in `test_logging.py` is run on a modified version of `LOGGING.data`. To resolve this I have changed the test cleanup from being performed after each test to before each test. 

The running of the file is generally unaffected; instead of having test 1, clean-up, test 2, clean-up, test 3, clean-up... we have clean-up, test 1, clean-up, test 2, clean-up, test 3... .

Alternative 1: We could have the clean-up be run after the tests automatically and manually enact it before the first test - the result would be virtually the same.

Alternative 2: Introducing data cleanup to the other test files that modify `LOGGING.data`.

I don't see any detriment to changing when the clean-up is introduced, be it before or after tests are called, other than feeling a bit counter-intuitive. However if we are against cleaning the data beforehand then I would vouch for alternative 2 as it allows for consistent implementation across the test suites and prevents this problem from happening in other tests in future. 

Would love to hear opinions from reviewers on which approach to go with.

Reply:
https://docs.pytest.org/en/6.2.x/fixture.html#teardown-cleanup-aka-fixture-finalization is a good read. I think it makes more sense to have a teardown process which clean up the test afterward.

Just to add more context, this is what happening now
`pytest tests/framework/project/test_logging.py  tests/framework/test_startup.py` - pass
 `pytest  tests/framework/test_startup.py  tests/framework/project/test_logging.py ` - fail

As test should be isolated, the problem here is then `test_startup.py` isn't clean up properly and we should update this file instead

Reply:
@AhdraMeraliQB I 100% agree with @noklam here. The state after running a test should be identical to what is was before running the test. Hence responsibility for clear up belongs in the teardown process rather than in the setup process. So alternative 2 is definitely the right one to go for here. If it means editing tests that already exist that is fine - it just means that those tests were not correctly doing teardown which they should have been doing before.